lib/wallet/src/Cardano/Wallet/DB/Store/DeltaUTxO/Model.hs

```diff
@@ -0,0 +1,164 @@
+            | At slot <= tip = noop
+            | otherwise =
+                UTxOHistory
+                    { history = history <> received delta
+                    , creationSlots =
+                        cleanUpSlots $
+                            Map.insert
+                                (At slot)
+                                ( dom (received delta)
+                                    `Set.difference` dom history
+                                )
+                                creationSlots
+                    , spentSlots =
+                        cleanUpSlots $
+                            spentSlots
+                                <> Map.fromList
+                                    [
+                                        ( slot
+                                        , excluded delta
+                                            `Set.intersection` dom history
+                                            `Set.difference` fold spentSlots
+                                        )
+                                    ]
+                    , tip = At slot
+                    , finality = finality
+                    }
+    apply
+        (Rollback slot)
+        noop@UTxOHistory{history, spentSlots, creationSlots, tip, finality}
+            | slot >= tip = noop
+            | otherwise =
+                let slot' = maybe slot (max slot) finality
+                    (leftCreationSlots, rolledSlots) =
+                        Map.spanAntitone (<= slot') creationSlots
+                    rolledTxIns = fold rolledSlots
+                 in UTxOHistory
+                        { history = history `excluding` rolledTxIns
+                        , spentSlots = case slot' of
+                            Origin -> mempty
+                            At slot'' ->
+                                Map.takeWhileAntitone (<= slot'') spentSlots
+                        , creationSlots = leftCreationSlots
+                        , tip = slot'
+                        , finality = finality
+                        }
+    apply
+        (Prune newFinality)
+        noop@UTxOHistory{history, spentSlots, creationSlots, tip, finality}
+            | checkFinalityIsOld finality = noop
+            | otherwise =
+                let newFinality' = min newFinality tip
+                    (prunedSlots, leftSlots) = case newFinality' of
+                        Origin -> (spentSlots, mempty)
+                        At slotNewFinality' ->
+                            Map.spanAntitone (<= slotNewFinality') spentSlots
+                    prunedTxIns = fold prunedSlots
+                 in UTxOHistory
+                        { history = history `excluding` prunedTxIns
+                        , spentSlots = leftSlots
+                        , creationSlots =
+                            cleanUpSlots $
+                                (`Set.difference` prunedTxIns) <$> creationSlots
```

Comment:
🤔 Hm. That also looks like an operation that takes O(n) time — this is the thing that we worried about, that pruning the end will also force a linear scan of the entire UTxO set. We may have to think about this again — but perhaps in a follow-up pull request.

Reply:
fixed